src/backend/columnar/sql/udfs/columnar_ensure_am_depends_catalog/latest.sql

```diff
@@ -0,0 +1,40 @@
+CREATE OR REPLACE FUNCTION citus_internal.columnar_ensure_am_depends_catalog()
+  RETURNS void
+  LANGUAGE plpgsql
+  SET search_path = pg_catalog
+AS $func$
+BEGIN
+  INSERT INTO pg_depend
+  SELECT -- Define a dependency edge from "columnar table access method" ..
+         'pg_am'::regclass::oid as classid,
+         (select oid from pg_am where amname = 'columnar') as objid,
+         0 as objsubid,
+         -- ... to each object that is registered to pg_class and that lives
+         -- in "columnar" schema. That contains catalog tables, indexes
+         -- created on them and the sequences created in "columnar" schema.
+         --
+         -- Given the possibility of user might have created their own objects
+         -- in columnar schema, we explicitly specify list of objects that we
+         -- are interested in.
+         'pg_class'::regclass::oid as refclassid,
+         columnar_schema_members::regclass::oid as refobjid,
+         0 as refobjsubid,
+         'n' as deptype
```

Comment:
I looked into the dependency types available. I'm not sure enough which one we want to use to suggest a change, but here's some things to consider:

`DEPENDENCY_NORMAL` seems like a reasonable solution, but not quite ideal. It would be a little weird to say "DROP TABLE columnar.options CASCADE", and have that work at all, much less cascade to drop all columnar tables. Of course only the extension owner could do that, so it kind of makes sense, but it's not necessarily the most expected behavior if it happens.

`DEPENDENCY_AUTO`, `DEPENDENCY_EXTENSION`, and `DEPENDENCY_PIN` are obviously not right.

`DEPENDENCY_INTERNAL`, where the the AM is the referenced object and the columnar metatables are the dependent objects (reverse of the normal dependency) seems almost like a good fit, because the metatables are more like an internal implementation detail of the AM. That would disallow dropping of the metatables (you'd need to drop the AM, which means you'd need to drop the extension). A brief experiment seems to show that this still dumps the metatable contents during `--binary-upgrade`. 

`DEPENDENCY_AUTO_EXTENSION` might work if we made every columnar table dependent on the extension. Not sure about this idea yet.

Reply:
> `DEPENDENCY_AUTO_EXTENSION` might work if we made every columnar table dependent on the extension. Not sure about this idea yet.

This comes somewhat close to what Citus does for distributed tables. We make all the distributed tables depend on the Citus extension to prevent the extension to be dropped without undistributing or dropping the distributed tables. It is especially important due to how the planner intercepts the reads and writes to distributed tables.

My hope would be that the tables already depend on the `AM`, which is part of the extension. So when dropping the Citus extension the AM needs to be dropped, which should result in an error as long as there are columnar tables. Unfortunately this does not apply the other way around where the `AM` is only used after the extension is fully instantiated during the upgrade path.

I haven't payed much attention to the dependency type. I have mostly thought about internal vs normal, and normal seemed the better approach.

@onurctirtir have you had a look at `DEPENDENCY_AUTO_EXTENSION`? Why would/wouldn't it work?

Reply:
I haven't look into that yet

Reply:
> @onurctirtir have you had a look at DEPENDENCY_AUTO_EXTENSION? Why would/wouldn't it work?

Today I tried that and it fixed the error that we get when dropping citus extension (see the ci failures), thank you @jeff-davis.

Then looked into pg code to see how it helps fixing regression tests, and my observation is as follows:
First of all, the codepath that we are interested in is: `​performMultipleDeletions -> findDependentObjects`​.
What I've seen is that `findDependentObjects` sorts depending objects by using ​`object_address_comparator`​​ when recursing through the dependency graph. And as a result, it deletes the objects in **descending order of their object addresses**. Moreover, the **primary sort key is the OID** itself:
```c
    /*
     * Primary sort key is OID descending.  Most of the time, this will result
     * in putting newer objects before older ones, which is likely to be the
     * right order to delete in.
     */
```

I think that explains how "adding DEPENDENCY_AUTO_EXTENSION edges from columnar tables to citus extension" fixes the regression test failures.
This is because, it's certain that any columnar table would have a greater OID both than citus extension itself and than any columnar metadata table.

Btw, even if it is not today's concern, given that PG15 would add support for `ALTER TABLE SET ACCESS METHOD`, the trick that we do here wouldn't help in that case. This is because, that would allow having columnar tables with smaller OID's from citus extension.
(Note that today this is not a problem for `alter_table_set_access_method` UDF, since it anyway creates the table from scratch.)

-----------

However, I'm still not very much convinced about why this is needed.
With the changes done in this pr, we already set-up following following dependency graph:

```
                  columnar_table
                       |
                       v
------------------- columnarAM ----> citus extension
|                     |                    ^  ^
|  -------------------                     |  |
|  |                                       |  |
|  --> columnar.storage_id_seq ----------  |  |
|                                             |
-----> columnar.stripe ------------------------
```

Shouldn't this graph already guarantee that the deletion of any columnar table will be done before dropping any columnar metadata table, or their toast rels. @jeff-davis @thanodnl ?

Reply:
"it's certain that any columnar table would have a greater OID both than citus extension itself"

That is false. I believe it will happen when you alter the access method. Even if not, OID wraparound can happen.

Reply:
"With the changes done in this pr...Shouldn't this graph already guarantee that the deletion of any columnar table will be done before dropping any columnar metadata table"

I believe that's correct. The question is: what is the best dependency type to use? Perhaps the question is academic, or perhaps it has some consequences that are worth exploring.

Reply:
> DEPENDENCY_INTERNAL, where the the AM is the referenced object and the columnar metatables are the dependent objects (reverse of the normal dependency) seems almost like a good fit, because the metatables are more like an internal implementation detail of the AM. That would disallow dropping of the metatables (you'd need to drop the AM, which means you'd need to drop the extension). A brief experiment seems to show that this still dumps the metatable contents during --binary-upgrade.

Usage of `DEPENDENCY_INTERNAL` seems like the right direction to me.
After reading more about the pg docs and the code comments, what I understood is that `internal` looks like something exactly what we are looking for.
As Jeff already mentioned, `obj_a --"i"--> obj_b` means a somewhat reverse dependency, where `obj_a` is part of the implementation of `obj_b`, meaning that `obj_a` cannot be dropped before dropping `obj_b`.

I think, and it seems that, this **should fix/fixes** the `pg_upgrade` issue.

Moreover, for the regression test failure, the code comment in `pg/dependency.c` states that:
```c
	/*
	 * The target object might be internally dependent on some other object
	 * (its "owner"), and/or be a member of an extension (also considered its
	 * owner).  If so, and if we aren't recursing from the owning object, we
	 * have to transform this deletion request into a deletion request of the
	 * owning object.  (We'll eventually recurse back to this object, but the
	 * owning object has to be visited first so it will be deleted after.) The
	 * way to find out about this is to scan the pg_depend entries that show
	 * what this object depends on.
	 */
```

In that sense, if we define internal dependencies from columnar metadata tables to columnarAM, then it will be guaranteed that the `columnarAM`, i.e.: the owning object, will be dropped after dropping `columnar.chunk`: 
```
toast_rel_columnar.chunk --i--> columnar.chunk --i--> columnarAM --i--> citus extension
```

However, taking dependency edges from columnar tables to columnarAM into account:
```
columnar_table ------------------------n----------------------
                                                             |
                                                             v 
toast_rel_columnar.chunk --i--> columnar.chunk --i--> columnarAM --i--> citus extension
```

, then DROP EXTENSION (or something similar) might still determine first dropping `toast_rel_columnar.chunk`, then `columnar_table` and then `citus extension`, which is the cause of the regression test failure.

**Even if this could be worked-around by applying some other tricks, I'm trying to understand if this dependency setup could also break `pg_upgrade` scenario.**

Reply:
> However, taking dependency edges from columnar tables to columnarAM into account:
> ```
> columnar_table ------------------------n----------------------
>                                                              |
>                                                              v 
> toast_rel_columnar.chunk --i--> columnar.chunk --i--> columnarAM --i--> citus extension
> ```
> , then DROP EXTENSION (or something similar) might still determine first dropping toast_rel_columnar.chunk, then columnar_table and then citus extension, which is the cause of the regression test failure.

From reading the code I was assuming this was a possibility, which seems wrong to me. Given deletion of `columnar_table` calls into `columnarAM` it should have removed all normal dependencies before starting to remove internal dependencies. If that rule was implemented it would also never remove `chunk`'s toast table.

Did you confirm from code analysis that it is indeed possible to remove normal dependencies after internal dependencies in the graph have already been removed? Are we able to reproduce this consistently?

Could we create a minimal viable example that exhibits this? Without relying on 200+ objects from the Citus extension? This would help discussions on `psql-hackers` as well as get advise from our committers.

Reply:
> Did you confirm from code analysis that it is indeed possible to remove normal dependencies after internal dependencies in the graph have already been removed?

Given that what `findDependentObjects` does is topological sorting in the high level, yes, I think I verified that it doesn't guarantee deletion order of the objects having no dependency relationship between them.

>  Are we able to reproduce this consistently?

Yes, the `drop extension` test in our regression test suite consistently reproduces the issue, but I'm not sure if it's something that we can share with `psql-hackers` directly, but sure, I can try to come up with a simpler example later.

Reply:
(nvm my comment, it doesn't even seem to be fixing the `pg_upgrade` issue)

Reply:
@jeff-davis With regards to `DROP TABLE columnar.options CASCADE;`:

The options table, as well as any other object from the extension, has a `DEPENDENCY_EXTENSION` dependency in pg_depend. When trying to `DROP CASCADE` the `columnar.options` I get the following error, both before and after the fix:

```
[local] nilsdijk@postgres:9700-26499=# DROP TABLE columnar.options CASCADE;
ERROR:  cannot drop table columnar.options because extension citus requires it
HINT:  You can drop extension citus instead.
Time: 6.238 ms
```

Given other dependency lines gave problems with actually dropping the extension I am inclined to keep this added dependency on `DEPENDENCY_NORMAL`.